### PR TITLE
Ide act synthesis

### LINKED ISSUE #603: IDE act: Get the Tutorial details from the course configuration file

### LINKED ISSUE #604: IDE act: create UI and connect to framework

### PR DESCRIPTION
**Pull request by OlliKiljunen**
# Description of the PR

+ Closes #604, closes #603. 

+ To see it in action, you have to use modified course config file. Download [o1_course_config.json](https://grader.cs.aalto.fi/static/O1_2020/projects/o1_course_config.json) to an empty dir in your machine. Add `tutorials` part to it (this turns GoodStuff exercise into a tutorial, for testing purposes):
```
{
  [...],
  "tutorials": {
    "24496": {
      "tasks": [
        {
          "instruction": "Find and open the file called CategoryDisplayWindow.scala",
          "component": "projectTree",
          "info": "You can see your project's files here.",
          "action": "openEditor",
          "actionArguments": {
            "filepath": "GoodStuff/o1/goodstuff/gui/CategoryDisplayWindow.scala",
          }
        },
        {
          "instruction": "Now, find and open the file called Experience.scala",
          "component": "projectTree",
          "info": "You can still see your project's files here.",
          "action": "openEditor",
          "actionArguments": {
            "filepath": "GoodStuff/o1/goodstuff/Experience.scala",
          }
        },
      ]
    }
  }
}
```
Start a server in the dir you stored the modified file:
```
python3 -m http.server 8000
```
Now you can use `http://localhost:8000/o1_course_config.json` as course config file URL. 

# Testing
<table>
  <tr>
    <th>unit</th>
    <th>integration</th>
    <th>e2e</th>
    <th>manual</th>
  </tr>
  <tr>
    <td>
      <ul>
        <li>- [ ] new <b>unit</b> tests created</li>
        <li>- [ ] all <b>unit</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>integration</b> tests created</li>
        <li>- [ ] all <b>integration</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>e2e</b> tests created</li>
        <li>- [ ] all <b>e2e</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] <b>manual</b> testing went well</li>
      </ul>
    </td>
  </tr>
</table>  
  
# Have you updated the [TESTING.md](https://github.com/Aalto-LeTech/intellij-plugin/blob/master/TESTING.md) or other relevant documentation on _your_ branch?

- [ ] Yes.
- [ ] Not yet. I will do it next.
- [x] Not relevant.

### CODE REVIEW

**Comment by OlliKiljunen** [comment]
If this is for any help:

Tutorials

**Review by stellatsv — APPROVED** [review]
I couldn't find anything that would need to be changed. Impeccable!

**Review by Taikelenn — APPROVED** [review]

File: src/main/java/fi/aalto/cs/apluscourses/presentation/ideactivities/TaskViewModel.java [hunk 1]
```diff
@@ -1,6 +1,17 @@
...
+
+  private final @NotNull Task task;
+
+  public TaskViewModel(@NotNull Task task) {
+    this.task = task;
+  }
+
+  public @NotNull String getTaskDescription() {
+    return "<html>" + task.getAction() + "</html>";
+  }
```
[the remark this answers is not in the document]
Comment (OlliKiljunen) [reply src/main/java/fi/aalto/cs/apluscourses/presentation/ideactivities/TaskViewModel.java:16]: It isn't. I'll remove it.

**Review by OlliKiljunen — COMMENTED** [review]
Thanks for reviews; I made changes per your suggestions.

File: src/main/java/fi/aalto/cs/apluscourses/utils/JsonUtil.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.IntFunction;
+import org.jetbrains.annotations.NotNull;
+import org.json.JSONArray;
+import org.json.JSONObject;
+
+public class JsonUtil {
```
[the remark this answers is not in the document]
Comment (OlliKiljunen) [reply src/main/java/fi/aalto/cs/apluscourses/utils/JsonUtil.java:12]: > the effort might be disproportional to the benefits

I agree plus I don't like this class very much; I wish we could replace it with something more sensible.

**Pull request merged**